internal/pkg/cosign/sign.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cosign
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"encoding/base64"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/pkg/errors"
+	"github.com/sigstore/cosign/cmd/cosign/cli/options"
+	"github.com/sigstore/cosign/pkg/oci"
+	"github.com/sigstore/cosign/pkg/oci/mutate"
+	ociremote "github.com/sigstore/cosign/pkg/oci/remote"
+	"github.com/sigstore/cosign/pkg/oci/static"
+	"github.com/sigstore/sigstore/pkg/signature"
+	signatureoptions "github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+type SigningRequest struct {
+	SignaturePayload []byte
+	SignedEntity     oci.SignedEntity
+}
+
+type SigningResults struct {
+	SignedPayload []byte
+	Signature     []byte
+	Pub           crypto.PublicKey
+	Cert, Chain   []byte
+	Bundle        *oci.Bundle
+
+	OCISignature oci.Signature
+	SignedEntity oci.SignedEntity
+}
+
+type Signer interface {
+	Sign(context.Context, *SigningRequest) (*SigningResults, error)
+}
+
+// PayloadSigner implements `Signer`
+type PayloadSigner struct {
+	PayloadSigner         signature.Signer
+	PayloadSignerOpts     []signature.SignOption
+	PublicKeyProviderOpts []signature.PublicKeyOption
+}
+
+// Sign uses the PayloadSigner to sign the requested payload, then returns the signature, the public key associated with it, the signed payload
+func (ps *PayloadSigner) Sign(ctx context.Context, req *SigningRequest) (*SigningResults, error) {
+	sOpts := []signature.SignOption{signatureoptions.WithContext(ctx)}
+	sOpts = append(sOpts, ps.PayloadSignerOpts...)
+	sig, err := ps.PayloadSigner.SignMessage(bytes.NewReader(req.SignaturePayload), sOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	pkOpts := []signature.PublicKeyOption{signatureoptions.WithContext(ctx)}
+	pkOpts = append(pkOpts, ps.PublicKeyProviderOpts...)
+	pk, err := ps.PayloadSigner.PublicKey(pkOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SigningResults{
+		Signature:     sig,
+		SignedPayload: req.SignaturePayload,
+		SignedEntity:  req.SignedEntity,
```

Comment:
This seems to be the only use of `req.SignedEntity`.  What are our expectations of the Payload?

Reply:
I was thinking a prior step might want to modify the `req.SignedEntity`, which is immutable, so I wanted to return the actual Entity in the results.

Our expectations for payload are to be able to upload it to Rekor, but we can get that from the `oci.SignedEntity`